services/code_hosting/git_hub_service.rb

```diff
@@ -1,118 +1,352 @@
+    # @return [String] HTML URL for the given pull request query.
+    def pull_requests(branches: nil, state: "open")
+      all_open_pull_requests = client.pull_requests(self.project.repo_config.full_name, state: state)
 
       # if no specific branch, return all open prs
-      return all_open_pull_requests if branches.nil? || branches.count == 0
-
-      branch_set = branches.to_set
-      all_open_pull_requests_on_branch = all_open_pull_requests.select { |pull_request| branch_set.include?(pull_request.head.ref) }
+      return all_open_pull_requests.map(&:html_url) if branches&.count == 0
 
+      pull_requests_on_branch = all_open_pull_requests.select { |pull_request| branches.include?(pull_request.base.ref) }
       # we want only the PRs whose latest commit was to one of the branches passed in
-      logger.debug("Returning all open prs from: #{repo_full_name}, branches: #{branches}, pr count: #{all_open_pull_requests.count}")
-      return all_open_pull_requests_on_branch
+      logger.debug("Returning all open prs from: #{self.project.repo_config.full_name}, branches: #{branches}, pr count: #{pull_requests_on_branch.count}")
+      return pull_requests_on_branch.map(&:html_url)
+    end
+
+    # Retrieve all commits' sha from a given repo_url and branch.
+    # @param [String] branch
+    # @return [Array<String>] List of SHA for a given branch and repo.
+    def all_commits_sha_for_branch(branch: nil)
+      return client.commits(self.project.repo_config.full_name, branch).map(&:sha)
     end
 
-    # returns only the most recent commit_sha for every open pr
-    # branches should be nil if you want all branches to be considered
-    def last_commit_sha_for_all_open_pull_requests(repo_full_name: nil, branches: nil)
-      return self.open_pull_requests(repo_full_name: repo_full_name, branches: branches).map { |pull_request| pull_request.head.sha }
+    # Retrieve the list of sha for a given pull request number.
+    # @param [Integer] number, of the pull request.
+    # @return [Array<String>] Array of SHA for the given pull request.
+    def commits_sha_from_pull_request(number: nil)
+      client.pull_commits(self.project.repo_config.full_name, number).map(&:sha)
+    end
+
+    # Returns the last sha for every pull request that targets a list of branches (or all if not given) and status (or open if not given)
+    # @param [Array<String>, nil] branches, Either an array of target branches names or nil.
+    # @param [String] state, Either open, closed, or all to filter by state. Default: open.
+    # @return [Array<String>] Array of the last commit SHA for the given repo and state.
+    def last_commit_sha_for_pull_requests(branches: nil, state: "open")
+      pull_requests_urls = self.pull_requests(branches: branches, state: state)
+      numbers = pull_requests_urls.map { |url| URI.parse(url) }.map(&:path).collect { |path| path.split("/").last }
```

Comment:
`numbers = pull_requests_urls.map { |url| URI.parse(url) }.map(&:path).collect { |path| path.split("/").last }`
`numbers = pull_requests_urls.map { |url| URI.parse(url).path.split("/").last }`

Reply:
When you go too far with functional programming

Reply:
haha, I love the functional aspect of it, but I'm like 99.99% sure ruby doesn't know how to optimize what you're doing there so I'm guessing the VM ends up creating like 1,000,000 temporary objects.